pkg/auth/authorizer/union/union.go

```diff
@@ -30,16 +30,25 @@ func New(authorizationHandlers ...authorizer.Authorizer) authorizer.Authorizer {
 }
 
 // Authorizes against a chain of authorizer.Authorizer objects and returns nil if successful and returns error if unsuccessful
-func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) error {
-	var errlist []error
+func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) (bool, string, error) {
+	var (
+		errlist []error
+		reason  string
+	)
 	for _, currAuthzHandler := range authzHandler {
-		err := currAuthzHandler.Authorize(a)
+		authorized, currentReason, err := currAuthzHandler.Authorize(a)
+		if currentReason != "" {
+			reason = currentReason
+		}
 		if err != nil {
 			errlist = append(errlist, err)
 			continue
 		}
-		return nil
+		if !authorized {
+			continue
+		}
+		return true, "", nil
 	}
 
-	return utilerrors.NewAggregate(errlist)
+	return false, reason, utilerrors.NewAggregate(errlist)
```

Comment:
return a joined list of reasons, maybe something like `strings.Join(reasonlist, "\n")`